File: battleship/static/engine/js/game.js
const rivalField = document.querySelector('.battlefield_rival #rival');
const selfField = document.querySelector('.battlefield_self #self');
const CLASS_MISS_CELL = 'battle-cell-miss';
const CLASS_MISSING = 'missing';
const CLASS_HIT_CELL = 'battle-cell-hit';
const btnRandomPlacement = document.getElementById('btn-random-placement');
const btnCleanSheet = document.getElementById('btn-clean-sheet');
const btnStartGamePC = document.getElementById('btn-start-game-pc');

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));
const host = window.location.host;
const url = 'http://' + host;
const url_ws = 'ws://' + host + '/ws/game_settings/';

let isGame = false;
let gameSocket = new WebSocket(url_ws);

gameSocket.onclose = closeWS;
gameSocket.onmessage = getMessageWS;
btnStartGamePC.onpointerup = startGamePC;
btnRandomPlacement.onpointerup = getRandomPlacement;
btnCleanSheet.onpointerup = cleanSheet;

async function closeWS () {
    while (gameSocket.readyState !== 1) {
        gameSocket = new WebSocket(url_ws);
        await sleep(3000);
    };
    gameSocket.onmessage = getMessageWS;
    gameSocket.onclose = closeWS;

}

function getMessageWS (msg) {
    let msgJSON = JSON.parse(msg.data);
    if (msgJSON.action === 'random_placement') {
        if (msgJSON.placement === null) return;
        putOnField(msgJSON);
    }
    else if (msgJSON.action === 'started_game') {

        createStatistcsDiv(msgJSON.body.kit_ships)
        if (msgJSON.body.turn_move === 'rival') {
            rivalField.onpointerup = null;
        }
        else {
            rivalField.onpointerup = makeMove;
        };
    }
    else if (msgJSON.action === 'return_motion') {
        if (msgJSON.body.turn_move === 'rival') {
            rivalField.onpointerup = null;
            rivalField.classList.add('field-disabled')
        }
        else {
            rivalField.onpointerup = makeMove;
            rivalField.classList.remove('field-disabled')
        }
        showResponseShot(msgJSON.body)
    }
};

function createDivEndGame () {
    let divEndGame = document.createElement('div');
    let p = document.createElement('p');
    let but = document.createElement('button');
    but.setAttribute('type', 'button');
    but.classList.add('custom-btn');
    but.classList.add('custom-btn-small');
    but.classList.add('btn-active');
    but.textContent = 'Сыграть еще раз';
    but.onpointerup = () => {
        window.location.href = url;
    };
    divEndGame.id = 'end-game';
    divEndGame.appendChild(p);
    divEndGame.appendChild(but);
    return divEndGame
}


function createDivShips (length, position, dataId = null, size=2) {
    let divShip = document.createElement('div');
    divShip.classList.add('ship-box');
    divShip.setAttribute('data-location', position);
    divShip.setAttribute('data-length', length);
    if (dataId !== null) {
        divShip.setAttribute('data-id', dataId);
    };

    divShip.style = `z-index: 1;`
    if (Number(length) === 1) {
        divShip.style.cssText += `width: ${size}.1rem;
                                height: ${size}.1rem;`;
    };
    if (position ===  'h') {
        divShip.style.cssText += `width: ${Number(length)*size}.1rem;
                                height: ${size}.1rem;`;
    }
    else {
        divShip.style.cssText += `width: ${size}.1rem;
                                height: ${Number(length)*size}.1rem;`;
    };
    return divShip
};

function createStatistcsDiv (kit_ships) {
    let divStatistics = document.createElement('div');
    divStatistics.id = 'statistics-ships';
    for (const[shipId, ship] of Object.entries(kit_ships)) {
        let divShip = createDivShips(ship.length, 'h', shipId, 1);
        //setup position:static and cursor default
        divShip.classList.add('port');
        divShip.classList.add('ui-draggable');
        divStatistics.appendChild(divShip);
    };
    let firstChild = document.querySelector('.block-init').firstChild;
    document.querySelector('.block-init').insertBefore(divStatistics, firstChild);
};

function putOnField (msgJSON) {
    matrix = msgJSON.placement;
    cleanField(selfField);

    let idShips = [];
    for (let [y, row] of matrix.entries()) {
        for (let [x, cell] of row.entries()) {
            if (typeof cell === 'string') {
                let ship = cell.split('-');
                let cellField = selfField.querySelector(`
                                td[data-x="${x}"][data-y="${y}"]`);
                cellField.classList.add(CLASS_BUSY_CELL);
                cellField.classList.remove(CLASS_EMPTY_CELL);
                if (idShips.includes(ship[2])) {
                    continue
                };
                let divShip = createDivShips(ship[0], ship[1], ship[2]);

                cellField.appendChild(divShip);
                idShips.push(ship[2]);
            }
        };
    };
    wrapperInitShips.style.display = 'none';
    removeBlockInitShips();
    Array.from(document.getElementsByClassName('ship-box')).forEach((elem) => {
        elem.onpointerdown = dragShip;
    });
};

function removeBlockInitShips () {
    let initShips = document.querySelector('.init-ships');
    if (initShips !== null) {
        document.querySelector('.init-ships').remove();
    }
};

function cleanField (field) {
    Array.from(field.getElementsByClassName('ship-box')).forEach((elem) => {
        elem.remove();
    });
    Array.from(field.querySelectorAll('td')).forEach((elem) => {
        elem.setAttribute("class", "");
        elem.classList.add(CLASS_EMPTY_CELL);
    });
};

function cleanSheet () {
    cleanField(selfField);
    let firstChild = document.querySelector('.block-init').firstChild;
    document.querySelector('.block-init').insertBefore(wrapperInitShips, firstChild);
    wrapperInitShips.style.display = '';
    wrapperInitShips.innerHTML = htmlInitShips;
    Array.from(document.getElementsByClassName('ship-box')).forEach((elem) => {
        elem.onpointerdown = dragShip;
    });
};


function getRandomPlacement () {
    if (isGame === true) return;
    cleanField(selfField);
    gameSocket.send(JSON.stringify({'action': 'get_random_placement'}));
};

function startGamePC (event, DefcountShip = 10,
                      ships = [4, 3, 3, 2, 2, 2, 1, 1, 1, 1]) {
    if (isGame === true) return;
    let countShips = 0;
    for (elem of document.getElementsByClassName('ship-box')) {
        countShips++;
        if (elem.parentElement.tagName !== 'TD') {
            alert('Выставьте все корабли на поле боя');
            return;
        };
    };
    if (countShips !== DefcountShip) {
        getRandomPlacement();
    };

    switchBtnActive();
    let matrix = getMatrix(selfField.children);
    gameSocket.send(JSON.stringify({"action": "start_game_with_pc",
                                    "matrix": matrix,
                                    "kit_ships": ships,
        })
    );
};

function switchBtnActive () {
    // Отключаем или включаем все кнопки и движение кораблей
    if (isGame === false) {
        isGame = true;
        for (elem of selfField.getElementsByClassName('ship-box')) {
            elem.style.cursor = 'default'
        };
        btnCleanSheet.onpointerup = null;
        btnRandomPlacement.onpointerup = null;
        btnStartGamePC.onpointerup = null;
        btnCleanSheet.classList.add('btn-disabled');
        btnRandomPlacement.classList.add('btn-disabled');
        btnStartGamePC.classList.add('btn-disabled');
        btnCleanSheet.classList.remove('btn-active');
        btnRandomPlacement.classList.remove('btn-active');
        btnStartGamePC.classList.remove('btn-active');
    }
    else {
        isGame = false;
        btnStartGamePC.onpointerup = startGamePC;
        btnRandomPlacement.onpointerup = getRandomPlacement;
        btnCleanSheet.onpointerup = cleanSheet;

        btnCleanSheet.classList.remove('btn-disabled');
        btnRandomPlacement.classList.remove('btn-disabled');
        btnStartGamePC.classList.remove('btn-disabled');
        btnCleanSheet.classList.add('btn-active');
        btnRandomPlacement.classList.add('btn-active');
        btnStartGamePC.classList.add('btn-active');
    }

}

function getMatrix (field) {
    let matrix = [];
    for (let y = 0; y < field.length; y++) {
        matrix[y] = [];
        for (let x = 0; x < field[0].children.length; x++) {
            matrix[y][x] = 0;
        };
    };
    for (let [fieldY, row] of Array.from(field).entries()) {
        for (let [fieldX, cell] of Array.from(row.children).entries()) {
            if (cell.classList.contains(CLASS_BUSY_CELL)) {
                let ship = cell.querySelector('.ship-box')
                if (ship) {
                    let length = ship.getAttribute('data-length');
                    let location = ship.getAttribute('data-location');
                    let dataId = ship.getAttribute('data-id');
                    dataShip = `${length}-${location}-${dataId}`;

                    let _coords = getCoordsShipWithNeighbour(fieldX, fieldY, location, length);
                    coordsShip = _coords['ship'];
                    coordNeighbour = _coords['neighbour'];
                    for (let coord of coordsShip) {
                        matrix[coord[1]].splice([coord[0]], 1, dataShip)
                    }
                };
            }
        };
    };
    return matrix
}


function makeMove (event) {
    if (isGame === false) {return};
    let field = event.target
    if (field.style.disabled === true ||
        field.tagName === 'SPAN'
    ) {return};
    let fieldX = Number(field.getAttribute('data-x'));
    let fieldY = Number(field.getAttribute('data-y'));
    gameSocket.send(JSON.stringify({'action': 'move_player',
                                    'move': [fieldX, fieldY],
    }));
};

function showResponseShot (body) {
    let status_cell = body.status_cell;
    let isSunk = body.sunk;
    let fieldX = body.x;
    let fieldY = body.y;
    let turn = body.is_it_move;
    let field
    let cell = null;

    if (turn === 'rival') {
        cell = selfField.querySelector(`td[data-x="${fieldX}"][data-y="${fieldY}"]`);
        field = selfField;
    }
    else {
        cell = rivalField.querySelector(`td[data-x="${fieldX}"][data-y="${fieldY}"]`);
        field = rivalField;
    };

    let cellLastShot = field.querySelector('.last-shot');
    if (cellLastShot !== null) {
        cellLastShot.classList.remove('last-shot');
    };
    cell.classList.add('last-shot');

    if (status_cell === null) {return};
    if (status_cell === 'hit') {
        cell.classList.remove(CLASS_EMPTY_CELL);
        cell.classList.add(CLASS_HIT_CELL);
        cell.style.disabled = true;
        let missAuto = [];
        if (isSunk === true) {
            if (turn === 'player') {
                let ship = body.ship;
                let startCoordShip = field.querySelector(`td[data-x="${ship.start_coord[0]}"][data-y="${
                                                                                ship.start_coord[1]}"]`);
                let divShip = createDivShips(ship.length, ship.position);
                startCoordShip.appendChild(divShip);
                updateStatisticsShips(ship.length)
            }
            missAuto = body.impossible_fields;
        }
        else {
            missAuto = [[fieldX+1, fieldY+1], [fieldX-1, fieldY-1], [fieldX-1, fieldY+1], [fieldX+1, fieldY-1]]
        }

        missAuto.forEach((coord) => {
            if (checkBounds(coord[0], coord[1]) === true) {
                let missField = field.querySelector(`.${CLASS_EMPTY_CELL}[data-x="${
                                            coord[0]}"][data-y="${
                                            coord[1]}"]`);
                if (missField === null) {return};
                missField.classList.remove(CLASS_EMPTY_CELL);
                missField.classList.add(CLASS_MISS_CELL);
                missField.style.disabled = true;
                missField.querySelector('span').classList.add(CLASS_MISSING)
            };
        });
        if (body.is_game === false) {
            rivalField.onpointerup = null;
            let txtWinner
            if (body.winner === 'rival') {
                txtWinner = 'Вы проиграли. Сыграть еще раз?'
            }
            else {
                txtWinner = 'Вы победили. Сыграть еще раз?'
            }
            let divEndGame = createDivEndGame();
            divEndGame.firstChild.textContent = txtWinner;
            document.body.appendChild(divEndGame);
        }
    }
    else if (status_cell === 'miss') {
        cell.classList.remove(CLASS_EMPTY_CELL);
        cell.classList.add(CLASS_MISS_CELL)
        cell.style.disabled = true;
        cell.querySelector('span').classList.add(CLASS_MISSING)
    };
};

function updateStatisticsShips (shipLength) {
    let divStatisticsShips = document.getElementById('statistics-ships');
    for (ship of divStatisticsShips.children) {
        if (shipLength == Number(ship.getAttribute('data-length')) &
            ship.classList.contains('statistics-ship-hit') === false) {
                ship.classList.add('statistics-ship-hit');
                return
            }
    }
}

File: battleship/engine/services/game.py
import asyncio
import json
import random
from dataclasses import asdict
from typing import TYPE_CHECKING, Any, Literal

from engine.types.response import BodyResponse, Response, ShipDict

from .field import fill_auto_miss, get_possible_hit_move, get_random_placement
from .ship import Ship

if TYPE_CHECKING:
    from engine.consumers import GameSettingsConsumer


class GameWithPC:
    def __init__(self,
                 consumer: 'GameSettingsConsumer',
                 player_matrix: list[list[int | str]]) -> None:
        self.consumer = consumer
        self.player_matrix = player_matrix
        self.pc_matrix = get_random_placement()
        self.kit_ship_player: dict[str, Ship] = self._get_kit_ship_from_matrix(
                                                                self.player_matrix)
        self.kit_ship_pc: dict[str, Ship] = self._get_kit_ship_from_matrix(
                                                                self.pc_matrix)
        self.turn_move: Literal['player', 'rival'] = random.choice(['player', 'rival'])
        self.is_game = True

    def _get_kit_ship_from_matrix(self,
                                  matrix: list[list[int | str]]) -> dict[str, Ship]:
        ships: dict[str, Ship] = {}
        for y, row in enumerate(matrix):
            for x, cell in enumerate(row):
                if type(cell) is not str:
                    continue
                data_ship = cell.split('-')
                if ships.get(data_ship[2]) is None:
                    ships[data_ship[2]] = Ship(
                                            position=data_ship[1],  # type: ignore[arg-type]
                                            length=int(data_ship[0]),
                                            data_id=data_ship[2])
                ships[data_ship[2]].real_length += 1
                ships[data_ship[2]].coords.append([x, y])
        return ships

    def serialize_kit_ship(self,
                           kit_ships: dict[str, Ship]) -> dict[str, dict[str, Any]]:
        serialize_kit_ship = {}
        for ship_id, ship in sorted(kit_ships.items(),
                                    key=lambda item: item[1].length,
                                    reverse=True):
            serialize_kit_ship[ship_id] = {'length': ship.length,
                                           'position': ship.position
                                           }
        return serialize_kit_ship

    def get_matrix_for_move(self) -> tuple[list[list[int | str]],
                                           dict[str, Ship]]:
        if self.turn_move == 'player':
            matrix = self.pc_matrix
            kit_ships = self.kit_ship_pc
        else:
            matrix = self.player_matrix
            kit_ships = self.kit_ship_player
        return matrix, kit_ships

    def _get_possible_move(self, matrix: list[list[int | str]]) -> list[list[int]]:
        possible_move: list[list[int]] = []
        high_priority_cell: list[list[int]] = []
        for y, row in enumerate(matrix):
            for x, cell in enumerate(row):
                if cell == 2:
                    auto_hit = get_possible_hit_move(x, y)
                    for field in auto_hit:
                        if matrix[field[1]][field[0]] not in [1, 2]:
                            high_priority_cell.append(field)
                if cell not in [1, 2]:
                    possible_move.append([x, y])
        if len(high_priority_cell) > 0:
            return high_priority_cell
        return possible_move

    def _pass_turn(self):
        if self.turn_move == 'player':
            self.turn_move = 'rival'
        else:
            self.turn_move = 'player'

    async def start_game(self) -> None:
        await self.consumer.send(json.dumps({
                    'action': 'started_game',
                    'body': {
                        'is_game': self.is_game,
                        'turn_move': self.turn_move,
                        'kit_ships': self.serialize_kit_ship(self.kit_ship_pc)
                        }

                    }))
        if self.turn_move == 'rival':
            await self.make_move()

    async def make_move(self) -> None:
        await asyncio.sleep(0.5)
        pc_possible_move = self._get_possible_move(self.player_matrix)
        coord = random.choice(pc_possible_move)
        await self.proccess_move(coord[0], coord[1])

    async def proccess_move(self, x: int, y: int) -> None:
        if self.is_game is False:
            return None
        response = Response(
                            'return_motion',
                            BodyResponse(
                                is_game=self.is_game,
                                is_it_move=self.turn_move,
                                turn_move=self.turn_move,
                                status_cell='miss',
                                sunk=False,
                                x=x,
                                y=y,
                                )
                            )

        matrix, kit_ships = self.get_matrix_for_move()
        # for MYPY
        if response.body is None:
            return None

        # cell = matrix[y][x]
        if type(matrix[y][x]) is str:
            ship_id = matrix[y][x].split('-')[2]
            ship = kit_ships[ship_id]
            ship.damaged_coords.append([x, y])
            ship.check_sunk()

            matrix = fill_auto_miss(matrix, y, x)

            response.body.status_cell = 'hit'
            response.body.impossible_fields = ship.neighnour_fields
            response.body.sunk = ship.sunk
            if ship.sunk:
                for neighbour in ship.neighnour_fields:
                    matrix[neighbour[1]][neighbour[0]] = 1
                response.body.ship = ShipDict(
                    start_coord=ship.get_start_coord(),
                    length=ship.length,
                    position=ship.position
                )

                for boat in kit_ships.values():
                    if not boat.sunk:
                        break
                else:
                    self.is_game = False
                    response.body.is_game = self.is_game
                    response.body.winner = self.turn_move
        else:
            self._pass_turn()
            matrix[y][x] = 1

        response.body.turn_move = self.turn_move
        await self.consumer.send(json.dumps(asdict(response)))

        if self.turn_move == 'rival':
            await self.make_move()
